### ISSUE TITLE
Use marshmallow-jsonapi?

### ISSUE THREAD

**Issue by sloria** [4Catalyzer/flask-jsonapiview#13]
You might find the marshmallow-jsonapi package useful: https://github.com/marshmallow-code/marshmallow-jsonapi .

Nice work on this project; I definitely see a need for something like this in the Flask ecosystem.

**Comment by taion** [comment]
Thanks! I'd been meaning to take a look at some of your extensions like marshmallow-sqlalchemy but hadn't known about marshmallow-jsonapi until now. It'd be really cool if I could restrict this library to just the view-related functionality.

Do you know of any similar packages in the Flask ecosystem? I spent a long time looking for something DRF-like in the Flask ecosystem that offered a similar model that made it useful for building both CRUD resources and more bespoke API endpoints, and everything I found seemed to either be fully auto-generated resources with a few config options (like Flask-Restless), or extremely thin wrappers over `MethodView` (like Flask-RESTful).

It's pretty nerve-wracking to try to figure out a generic way to handle things like filtering and sorting, or to handle relationships at all.

**Comment by sloria** [comment]
Indeed, it would be great to have a library that includes a set of DRF-like generic views. I, too, have yet to find a package in the Flask ecosystem that meets the right level of abstraction. The closest thing I've seen is probably [Flask-Potion](https://potion.readthedocs.org/en/latest/), but I haven't looked much into it yet.

**Comment by sloria** [comment]
Another one worth mentioning is [restless](https://github.com/toastdriven/restless) (not to be confused with Flask-Restless).

**Comment by taion** [comment]
I looked at Flask-Potion a little bit - I felt like it moved a lot in the right direction by using class-based views, compared to something like Flask-Restless. restless didn't seem to offer that much about e.g. Flask-RESTful, though.

I guess the thing I thought I was missing from a lot of those was that I couldn't find any libraries that both had a concept of pagination (which the smaller frameworks don't), while not forcing me into a single specific style of pagination.

And then likewise for everything else like sorting, filtering, auth, &c.

**Comment by sloria** [comment]
Looks look you're making progress on those fronts (pagination, sorting, auth, etc.). Re: authn, I recommend taking a look at this: http://plope.com/pyramid_auth_design_api_postmortem for design ideas. For authz, http://plope.com/pyramid_auth_design_api_postmortem . [Flask-Asylum](https://github.com/mattupstate/flask-asylum) attempted to implement these ideas as a Flask extension, but it appears to be unmaintained ATM.

**Comment by taion** [comment]
I've mostly just been following what DRF has, with a few tweaks for what makes the most sense for our use case - ideally this should be enough to build something more generic, since I think the pain points I have also apply to some other more modern applications as well.

On the authN front, I think I have something in between what Pyramid has and DRF authN policies. I don't think API endpoints in general need to care about remembering or forgetting a user. DRF does things fairly well here, except they're tied to user objects from the ORM. In my case, I'm using a 3rd party (Auth0) for auth, and just pulling off JWTs from each request. My own APIs don't deal with session management at all, and the token won't necessarily even identify a user (v. just a set of permissions in the abstract), nor might I even have my own User model (I don't).

AuthZ is where things get a little hairier for me. Our system is multi-tenant, so being able to filter a SQLAlchemy query based on request credentials is fundamental to being able to operate efficiently. I think in DRF this is handled as a filter on the queryset, but IMO the authZ concern of "filter the query to just the relevant objects (e.g. those associated with the current tenant)" belongs with authZ, since it is part of authZ. And I think in general for multi-tenant systems, stock ACLs don't seem too expressive given the need to manage permissions separately per-tenant.

There's also a few other minor annoyances there for handling multi-tenant systems. Not only do I not want a user to be able to modify an object belonging to another tenant, I also don't want a user to be able to set the tenant on an object to something other than his or her own tenant. I can handle this as a deserialization concern and just ignore that field, but I feel like this is still better expressed as an authZ thing. I don't know. I think the API still allows for something ACL-based; I just wouldn't be using such a system.

Long story short, my hope is that by taking DRF-style APIs and making them a little more generic to handle quirks in my use case can let me end up with something that still works as well for typical use cases, but like Flask in general is better able to handle more atypical ones.

**Comment by taion** [comment]
Mostly a reminder to myself - I haven't actually forgotten about this and will do this when I have time.

**Comment by sloria** [comment]
Great; let me know if I can help in any way.

**Comment by taion** [comment]
Epic fail on my part...

I never realized that JSON API moved attributes into a nested `attributes` object instead of just being on top level with `type` and `id`.

I think I'm just going to drop JSON API and, uhh, find a better name for this library.

Sorry!

**Comment by sloria** [comment]
Fair enough; JSON API formatting seems to be outside the scope of this library. Too bad Flask-API is already taken.. =)

**Comment by taion** [comment]
Yeah - I'm going to drop all the bad attempts I've made at doing my own JSON API formatting - probably just keep the `data`/`meta` split because it's useful for communicating pagination data.

Probably makes more sense anyway - and if I do decide to go with JSON API, should be straightforward enough to plug in marshmallow-jsonapi after the fact.

And next time a spec hits 1.0, I'm actually going to read the whole thing instead of assuming I know what it says because I looked at the RC.
